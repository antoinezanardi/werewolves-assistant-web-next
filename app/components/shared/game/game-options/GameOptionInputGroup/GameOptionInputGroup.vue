<template>
  <div
    id="game-option-input-group-container"
  >
    <div
      id="game-option-input-group"
      class="flex gap-2 items-center justify-between"
    >
      <h3
        id="game-option-label"
        class="w-8/12"
      >
        <FontAwesomeIcon
          id="game-option-label-icon"
          class="me-2"
          :class="optionIconClass"
          :icon="optionIcon"
        />

        <span>
          {{ optionLabel }}
        </span>
      </h3>

      <div
        id="game-option-input"
        class="w-4/12"
      >
        <slot/>
      </div>
    </div>

    <p
      id="game-option-description"
      class="italic mt-3 text-gray-400"
    >
      {{ optionDescription }}
    </p>

    <PrimeVueDivider
      v-if="doesHaveBottomDivider"
      id="game-option-bottom-divider"
      class="!my-4"
    />
  </div>
</template>

<script setup lang="ts">
import { FontAwesomeIcon } from "@fortawesome/vue-fontawesome";
import type { GameOptionInputGroupProps } from "~/components/shared/game/game-options/GameOptionInputGroup/game-option-input-group.types";

withDefaults(defineProps<GameOptionInputGroupProps>(), { doesHaveBottomDivider: false });

defineSlots<{
  default: () => void;
}>();
</script>