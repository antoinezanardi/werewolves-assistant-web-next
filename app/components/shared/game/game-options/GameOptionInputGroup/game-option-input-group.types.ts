type GameOptionInputGroupProps = {
  optionLabel: string;
  optionIcon: string;
  optionIconClass?: string;
  optionDescription: string;
  doesHaveBottomDivider?: boolean;
};

export type { GameOptionInputGroupProps };