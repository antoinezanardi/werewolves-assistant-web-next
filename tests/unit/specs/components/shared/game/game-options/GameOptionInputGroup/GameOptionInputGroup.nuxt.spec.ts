import type { FontAwesomeIcon } from "@fortawesome/vue-fontawesome";
import type { mount } from "@vue/test-utils";
import type { ComponentMountingOptions } from "@vue/test-utils/dist/mount";
import type Divider from "primevue/divider";

import type { GameOptionInputGroupProps } from "~/components/shared/game/game-options/GameOptionInputGroup/game-option-input-group.types";
import GameOptionInputGroup from "~/components/shared/game/game-options/GameOptionInputGroup/GameOptionInputGroup.vue";
import { mountSuspendedComponent } from "@tests/unit/utils/helpers/mount.helpers";

describe("Game Option Input Group Component", () => {
  let wrapper: ReturnType<typeof mount<typeof GameOptionInputGroup>>;
  const defaultProps: GameOptionInputGroupProps = {
    optionLabel: "label",
    optionIcon: "check",
    optionIconClass: "text-success",
    optionDescription: "description",
  };

  async function mountGameOptionInputGroupComponent(options: ComponentMountingOptions<typeof GameOptionInputGroup> = {}):
  Promise<ReturnType<typeof mount<typeof GameOptionInputGroup>>> {
    return mountSuspendedComponent(GameOptionInputGroup, {
      props: defaultProps,
      ...options,
    });
  }

  beforeEach(async() => {
    wrapper = await mountGameOptionInputGroupComponent();
  });

  it("should match snapshot when rendered.", () => {
    expect(wrapper).toBeTruthy();
    expect(wrapper.html()).toMatchSnapshot();
  });

  describe("Label Icon", () => {
    it("should pass the icon name from props to the label icon when rendered.", () => {
      const labelIcon = wrapper.findComponent<typeof FontAwesomeIcon>("#game-option-label-icon");

      expect(labelIcon.props("icon")).toBe("check");
    });

    it("should pass the icon class from props to the label icon when rendered.", () => {
      const labelIcon = wrapper.findComponent<typeof FontAwesomeIcon>("#game-option-label-icon");

      expect(labelIcon.attributes("class")).toContain(defaultProps.optionIconClass);
    });
  });

  describe("Label Text", () => {
    it("should pass the label text from props to the label text when rendered.", () => {
      const labelText = wrapper.find<HTMLHeadingElement>("#game-option-label");

      expect(labelText.text()).toBe(defaultProps.optionLabel);
    });
  });

  describe("Description", () => {
    it("should pass the description text from props to the description when rendered.", () => {
      const description = wrapper.find<HTMLParagraphElement>("#game-option-description");

      expect(description.text()).toBe(defaultProps.optionDescription);
    });
  });

  describe("Bottom Divider", () => {
    it("should not render the bottom divider when doesHaveBottomDivider prop is not provided.", () => {
      const divider = wrapper.findComponent<typeof Divider>("#game-option-bottom-divider");

      expect(divider.exists()).toBeFalsy();
    });

    it("should render the bottom divider when doesHaveBottomDivider prop is provided as true.", async() => {
      wrapper = await mountGameOptionInputGroupComponent({
        props: {
          ...defaultProps,
          doesHaveBottomDivider: true,
        },
      });
      const divider = wrapper.findComponent<typeof Divider>("#game-option-bottom-divider");

      expect(divider.exists()).toBeTruthy();
    });
  });
});