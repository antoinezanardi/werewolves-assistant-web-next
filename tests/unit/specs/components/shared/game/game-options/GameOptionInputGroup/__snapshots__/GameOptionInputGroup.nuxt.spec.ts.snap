// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Game Option Input Group Component > should match snapshot when rendered. 1`] = `
"<div id="game-option-input-group-container">
  <div id="game-option-input-group" class="flex flex-col gap-2 items-center justify-between md:flex-row">
    <h3 id="game-option-label" class="flex gap-2 items-center md:text-left md:w-8/12 text-center w-full">
      <font-awesome-icon-stub icon="check" border="false" fixedwidth="false" flip="false" listitem="false" pulse="false" swapopacity="false" spin="false" symbol="false" inverse="false" bounce="false" shake="false" beat="false" fade="false" beatfade="false" flash="false" spinpulse="false" spinreverse="false" id="game-option-label-icon" class="text-center text-success"></font-awesome-icon-stub><span class="grow md:pe-0 pe-3 text-pretty">label</span>
    </h3>
    <div id="game-option-input" class="md:w-4/12 w-full"></div>
  </div>
  <p id="game-option-description" class="italic mt-3 text-gray-400">description</p>
  <!--v-if-->
</div>"
`;
